java-client/session/src/main/java/io/deephaven/client/impl/ObjectService.java

```diff
@@ -6,11 +6,123 @@
 import java.util.concurrent.CompletableFuture;
 
 public interface ObjectService {
+
+    /**
+     * A server object that supports fetching.
+     */
+    interface Fetchable extends ServerObject {
+        /**
+         * The type.
+         *
+         * @return the type
+         */
+        String type();
+
+        /**
+         * Fetches {@code this}. The resulting object is managed separately than {@code this}.
+         *
+         * @return the future
+         * @see Session#fetch(HasTypedTicket) for the lower-level interface
+         */
+        CompletableFuture<DataAndExports> fetch();
+    }
+
+    /**
+     * A server object that supports a bidirectional message stream.
+     */
+    interface Bidirectional extends ServerObject {
+        /**
+         * The type.
+         *
+         * @return the type
+         */
+        String type();
+
+        /**
+         * Opens a bidirectional message stream for {@code this}. The returned {@link DataAndExports} messages are
+         * managed separately than {@code this}.
+         *
+         * <p>
+         * This provides a generic stream feature for Deephaven instances to use to add arbitrary functionality. This
+         * API lets a client open a stream to a particular object on the server, to be mediated by a server side plugin.
+         * In theory this could effectively be used to "tunnel" a custom gRPC call, but in practice there are a few
+         * deliberate shortcomings that still make this possible, but not trivial.
+         *
+         * <p>
+         * Presently it is required that the server respond immediately, at least to acknowledge that the object was
+         * correctly contacted (as opposed to waiting for a pending ticket, or dealing with network lag, etc). This is a
+         * small (and possibly not required, but convenient) departure from a offering a gRPC stream (a server-streaming
+         * or bidi-streaming call need not send a message right away).
+         *
+         * @param stream the stream where the client will receive messages
+         * @return the stream where the client will send messages
+         * @see Session#messageStream(HasTypedTicket, MessageStream) for the lower-level interface
+         */
+        MessageStream<DataAndTypedTickets> messageStream(MessageStream<DataAndExports> stream);
```

Comment:
I missed your note about how this patch now includes a high level api, and if I'm reading the api properly, the `Bidirectional` type is returned by the high level api (so itself is part of the high level api). But it still returns a stream implementation that is dependent on high-level client code being able to transform tickets into objects (like tables, other plugins, etc)?

Reply:
`Fetchable` and `Bidirectional` are the high-level apis; the user's shouldn't need to know anything about transforming tickets. Both of these are client-managed exports, and both present `DataAndExports` to the users. The important part, also what I consider high-level, is `ServerObject` (`List<ServerObject> exports`); these are also client-managed exports, but provide users differentiation between: `TableObject`, `CustomObject extends Fetchable, Bidirectional`, and `UnknownObject`.